### PR TITLE
Jax bindings support

### PR DESCRIPTION
**Pull request by laserkelvin**
# ALCHEMI Toolkit-Ops Pull Request

## Description

This PR adds `jax` support to `nvalchemiops`.

- General use patterns for `jax` bindings is divergent from `torch` due to `jax.jit` requirements. Instead of calling `wp.launch`, the `warp` kernels are wrapped with `warp.jax_experimental.jax_kernel` and treated as if they are `jax` native functions. While some of the underlying functions (particularly overloads) are not part of the public API, users could technically call them directly and we may decide later if we want to expose them.
    - One big difference is the use of `estimate_cell_list_sizes` and `build_cell_lists`: these are incompatible with `jax.jit` because they rely on runtime values to allocate arrays and hence untraceable. If `jax.jit` is desirable **through** cell list operations, users are required to pass a `max_total_cells` to use static shaped arrays.
    - `jax.jit` smoke tests are included in the test suite on an individual basis. Known failure cases (e.g. like the one above) are marked with `xfail` so we return to them later.
- `jax` differentiation is not currently supported: the main reason being the structure of some of the underlying `warp` kernels are written with mutation: **multiple** output arrays are pre-allocated and modified, which is not supported by `jax_kernel(..., enable_backwards=True)` which handles output allocation by itself. If this support is required, we might need to look at refactoring how the `warp` kernels are structured.

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- Add `nvalchemops.jax` namespace
- For the most part, launchers are available and function end-to-end, with the caveat that some may not be `jax.jit` compatible for the aforementioned reasons.

## Testing

- [ ] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations?

## Checklist

- [ ] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [ ] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/jax/neighbors/batch_naive_dual_cutoff.py [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+        ):
+            shift_range_per_dimension, shift_offset, total_shifts = (
+                compute_naive_num_shifts(cell, cutoff2, pbc)  # Use larger cutoff
+            )
+
+        # Compute max_atoms_per_system if needed
+        if max_atoms_per_system is None:
+            # Compute from batch_ptr
+            atoms_per_system = batch_ptr[1:] - batch_ptr[:-1]
+            max_atoms_per_system = int(jnp.max(atoms_per_system))
```
Comment (zubatyuk) [inline nvalchemiops/jax/neighbors/batch_naive_dual_cutoff.py]: wrap in `try ... except (jax.errors.ConcretizationTypeError, jax.errors.TracerIntegerConversionError)`

File: nvalchemiops/jax/neighbors/__init__.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        case "batch_cell_list":
+            # NOTE: JAX batch_cell_list does not yet support half_fill/fill_value
+            # (unlike Torch). These parameters are silently ignored here.
+            # See JAX_FINAL.md for tracking.
+            return batch_cell_list(
+                positions,
+                cutoff,
+                cell,
+                pbc,
+                batch_idx,
```
Comment (zubatyuk) [inline nvalchemiops/jax/neighbors/__init__.py:324]: batch_ptr missing

File: nvalchemiops/jax/neighbors/batch_cell_list.py [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+    # Ensure pbc is bool with shape (num_systems, 3)
+    if pbc is not None:
+        pbc_bool = pbc.astype(jnp.bool_)
+    else:
+        pbc_bool = jnp.ones((num_systems, 3), dtype=jnp.bool_)
+
+    total_atoms = positions.shape[0]
+
+    # Allocate neighbor_matrix_shifts
+    neighbor_matrix_shifts = jnp.zeros(
```
Comment (zubatyuk) [inline nvalchemiops/jax/neighbors/batch_cell_list.py]: shoud the function accept pre-allocated buffers?

File: nvalchemiops/jax/neighbors/naive.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        fill_value = positions.shape[0]
+
+    if neighbor_matrix is None:
+        neighbor_matrix = jnp.full(
+            (positions.shape[0], max_neighbors),
+            fill_value,
+            dtype=jnp.int32,
+        )
+    else:
+        neighbor_matrix = neighbor_matrix.at[:].set(fill_value)
```
Comment (zubatyuk) [inline nvalchemiops/jax/neighbors/naive.py:256]: .set creates new array. XLA might optimize away the copy when array has no other references, no guarantee. Should the docstring be updated?

File: nvalchemiops/jax/neighbors/__init__.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            Pre-allocated tensor of shape (max_total_cells,) for number of atoms in each cell.
+            Can be provided to avoid reallocation for cell list construction.
+        cell_atom_start_indices : jax.Array, optional
+            Pre-allocated tensor of shape (max_total_cells,) for starting index in
+            cell_atom_list for each cell. Can be provided to avoid reallocation for
+            cell list construction.
+        cell_atom_list : jax.Array, optional
+            Pre-allocated tensor of shape (total_atoms,) for flattened list of atom
+            indices organized by cell. Can be provided to avoid reallocation for
+            cell list construction.
```
Comment (zubatyuk) [inline nvalchemiops/jax/neighbors/__init__.py:185]: Does not match signature for torch bindings.

File: nvalchemiops/jax/interactions/dispersion/_dftd3.py [hunk 6]
```diff
@@ -0,0 +1,979 @@
...
+
+    # Set fill_value if not provided
+    if fill_value is None:
+        fill_value = num_atoms
+
+    # Handle empty case
+    if num_atoms == 0:
+        num_systems = 1
+        if batch_idx is not None:
+            num_systems = int(jnp.max(batch_idx)) + 1
```
Comment (zubatyuk) [inline nvalchemiops/jax/interactions/dispersion/_dftd3.py]: wrap in try ... except (jax.errors.ConcretizationTypeError, jax.errors.TracerIntegerConversionError)
There are more occurrences in this file.

File: nvalchemiops/jax/neighbors/batch_naive_dual_cutoff.py (hunk 1 above)
Comment (laserkelvin) [reply nvalchemiops/jax/neighbors/batch_naive_dual_cutoff.py]: Done in d59f49d

File: nvalchemiops/jax/neighbors/__init__.py (hunk 2 above)
Comment (laserkelvin) [reply nvalchemiops/jax/neighbors/__init__.py:324]: Done in 2ca05d8

File: nvalchemiops/jax/neighbors/batch_cell_list.py (hunk 3 above)
Comment (laserkelvin) [reply nvalchemiops/jax/neighbors/batch_cell_list.py]: Done in 3c58722

File: nvalchemiops/jax/neighbors/__init__.py (hunk 5 above)
Comment (laserkelvin) [reply nvalchemiops/jax/neighbors/__init__.py:185]: Fixed in 6c3282f

File: nvalchemiops/jax/neighbors/naive.py (hunk 4 above)
Comment (laserkelvin) [reply nvalchemiops/jax/neighbors/naive.py:256]: Fixed in f730fed

**Comment by laserkelvin** [comment]
/ok to test [166a00f](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/24/commits/166a00f5b3b48f723c6006dca10c3dcc65a8d56c)

**Comment by laserkelvin** [comment]
/ok to test [8c13e68](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/24/commits/8c13e687cceb2236d69bd0247df54667b967f337)

**Comment by laserkelvin** [comment]
/ok to test [dfa335c](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/24/commits/dfa335cd908788689ea50ff8500d6c9f69df051b)
